pkg/storage/cacher.go

```diff
@@ -859,9 +859,27 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case !curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
 	}
+
+	// We need to ensure that if we put event X to the c.result, all
+	// previous events were already put into it before, no matter whether
+	// c.done is close or not.
+	// Thus we cannot simply select from c.done and c.result and this
+	// would give us non-determinism.
+	// At the same time, we don't want to block infinitely on putting
+	// to c.result, when c.done is already closed.
+
+	// This ensures that with c.done already close, we at most once go
+	// into the next select after this. With that, no matter which
+	// statement we choose there, we will deliver only consecutive
+	// events.
+	select {
+	case <-c.done:
+		return
+	default:
+	}
```

Comment:
As @xiang90 wrote, what you wrote can happen, but this is ok for us.
Basically, we only need to avoid situation when we from 2 consecutive events A, B, we don't deliver A, but then deliver B. (Either delivering both or neither of them is fine). 
If c.done will be satisfied after the above select, but before the one below, we don't know what will happen with A, but in the next round, we are sure that B won't be delivered. So we will be in the situation that we are fine with.